src/drawing-tool/components/runtime.tsx

```diff
@@ -140,7 +144,66 @@ export const Runtime: React.FC<IProps> = ({ authoredState, interactiveState, set
     }
   };
 
+  const handleUploadBtnClick = () => {
+    setUploadControlsVisible(true);
+  };
+  
+  const uploadFile = (fileOrUrl: File | string) => {
+    setUploadControlsVisible(false);
+    setUploadInProgress(true);
+    // Always copy image to S3 using Shutterbug, so it works even when the image disappears from the external location.
+    // Local image could be theoretically stored as dataSrc but it might be too big for Firestore storage that
+    // is used by ActivityPlayer. So, copying to S3 is a safer option.
+    let uploadPromise;
+    if (typeof fileOrUrl === "string") {
+      uploadPromise = copyImageToS3(fileOrUrl);
+    } else {
+      uploadPromise = copyLocalImageToS3(fileOrUrl);
+    }
+    uploadPromise
+      .then(url => {
+        handleSetInteractiveState({ userBackgroundImageUrl: url });
+      })
+      .catch((error) => {
+        window.alert(error);
```

Comment:
The `showModal("alert")` API is implemented on the LARA side (using existing jQuery alerts) and the API seems unlikely to change much, so it's stable in that sense. I don't think it has any existing clients at this point, but this would be a reasonable opportunity to use it. It would need to be implemented in the ActivityPlayer as well, however. Next time.